# Part 1. Equation of a Slime

How many late days are you using for this assignment?
0

In [105]:
pip install pandas


[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: C:\Users\raven\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [106]:
pip install scikit-learn

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: C:\Users\raven\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [107]:
# Imports section
import sklearn as skl
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.preprocessing import PolynomialFeatures, StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.model_selection import cross_val_score, ShuffleSplit, KFold
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn import svm
import pandas as pd

## 1. Loading the dataset

In [108]:
# Using pandas load the dataset
SD = pd.read_csv("./science_data_large.csv")
# Output the first 15 rows of the data
print(SD.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(SD.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [109]:
# Take the pandas dataset and split it into our features (X) and label (y)
x = SD[["Temperature °C", "Mols KCL"]]
y = SD["Size nm^3"]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
x_train, x_test, y_train, y_test = skl.model_selection.train_test_split(x, y, train_size=0.9, random_state=42)
print(x_train)
print(x_test)
print(y_train)
print(y_test)

     Temperature °C  Mols KCL
716              95       499
351             243       384
936              47       211
256             495       497
635              82       421
..              ...       ...
106             439        40
270             665       245
860             471       620
435             829       259
102             636       388

[900 rows x 2 columns]
     Temperature °C  Mols KCL
521             100       541
737             635       668
740             799       665
660             966       871
411             785       595
..              ...       ...
436             421       725
764             431       860
88              983       433
63              580       817
826              24        89

[100 rows x 2 columns]
716    103064.31430
351    193753.02860
936     21670.02857
256    505027.40000
635     75092.02857
           ...     
106     40433.71429
270    335545.00000
860    600674.85710
435    441286.60000
102    505469.25710
Name: Size n

## 3. Perform a Linear Regression

In [110]:
# Use sklearn to train a model on the training set
reg = LinearRegression().fit(X=x_train, y=y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample = pd.DataFrame()
sample["Temperature °C"] = [550]
sample["Mols KCL"] = [550]
print(reg.predict(sample))
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
print(reg.score(X=x_test, y=y_test))
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
print(reg.coef_)
print(reg.intercept_)

[634971.33433834]
0.8552472077276095
[ 866.14641337 1032.69506649]
-409391.47958340764


Write the linear equation of a slime: (example equation: $E = mc^2$)

t = temperature, m = Mols KCL, s = Size nm^3

$-409391.47958340764 + 866.14641337t + 1032.69506649m = s$

Report on score and explain meaning:

The score was 0.8552472077276095, which means that the ratio of the residual and total sum of squares is 0.1447527923, and the prediction is mostly correct. The higher the score, the better the model.

## 4. Use Cross Validation

In [111]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
cv = ShuffleSplit(n_splits=5, random_state=42)
scores = cross_val_score(reg, X=x, y=y, cv=cv)
print(scores)

[0.85524721 0.85830729 0.8654508  0.83855498 0.83002469]


Write findings here:

The cross validation scores are as follows:

[0.85524721 0.85830729 0.8654508  0.83855498 0.83002469]

These are the scores that the model got over the 5 shuffles of the data

## 5. Using Polynomial Regression

In [112]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
model = Pipeline([
    ("poly_features", PolynomialFeatures()),
    ("lin_reg", LinearRegression())
])
model.fit(X=x_train, y=y_train)

# Perform k-fold cross validation (as above)
kf = KFold(n_splits=5)
kf_scores = cross_val_score(model, X=x, y=y, cv=kf)
# Report on the metrics and output the resultant equation as you did in Part 3.
print(model.predict(sample))
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
print(model.score(X=x_test, y=y_test))
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
print(model.named_steps['lin_reg'].coef_)
print(model.named_steps["poly_features"].powers_)
print(model.named_steps['lin_reg'].intercept_)

[620242.85713165]
1.0
[ 0.00000000e+00  1.20000000e+01 -1.27197088e-07  1.26458843e-11
  2.00000000e+00  2.85714287e-02]
[[0 0]
 [1 0]
 [0 1]
 [2 0]
 [1 1]
 [0 2]]
2.0479375962167978e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

t = temperature, m = Mols KCL, s = Size nm^3

$2.0479375962167978*10^{-5} + 12t - 1.27197088*10^{-7}m + 1.26458843*10^{-11}t^2 + 2tm + 2.85714287*10^{-2}m^2 = s$

Report on the score and interpret:

The score is 1.0 meaning that the ratio of the residual and total sum of squares is 0, meaning that the residual sum of squares is 0, meaning the model perfectly matches the test data

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [113]:
# Load the dataset. Then train and evaluate the classification models.
CKD = pd.read_csv("./ckd_feature_subset.csv")
x = CKD.drop("Target_ckd", inplace=False, axis=1)
y = CKD["Target_ckd"]
x_train, x_test, y_train, y_test = skl.model_selection.train_test_split(x,y, train_size=0.9, random_state=42)

results = pd.DataFrame(index=["Logistic Regression", "Support Vector Machines", "k-Nearest Neighbors"])

### Logistic Regression

In [114]:
log_reg = LogisticRegression(random_state=42).fit(X=x_train, y=y_train)

log_kf_scores = cross_val_score(log_reg, X=x, y=y, cv=kf)

results.at["Logistic Regression", "scores"] = str(log_kf_scores)

results.at["Logistic Regression", "avg"] = log_kf_scores.mean()

results.at["Logistic Regression", "std"] = log_kf_scores.std()

### Support Vector Machines

In [115]:
svm_model = svm.SVC(random_state=42).fit(X=x_train, y=y_train)

svm_kf_scores = cross_val_score(svm_model, X=x, y=y, cv=kf)

results.at["Support Vector Machines", "scores"] = str(svm_kf_scores)

results.at["Support Vector Machines", "avg"] = svm_kf_scores.mean()

results.at["Support Vector Machines", "std"] = svm_kf_scores.std()


### k-Nearest Neighbors

In [116]:
model = Pipeline(
    steps=[("scaler", StandardScaler()), ("knn", KNeighborsClassifier())]
)

model.fit(X=x_train, y=y_train)

kn_kf_scores = cross_val_score(model, X=x, y=y, cv=kf)

results.at["k-Nearest Neighbors", "scores"] = str(kn_kf_scores)

results.at["k-Nearest Neighbors", "avg"] = kn_kf_scores.mean()

results.at["k-Nearest Neighbors", "std"] = kn_kf_scores.std()

## Results and Conclusion for Classification Experiments

In [117]:
print(results)

                                                                    scores  \
Logistic Regression      [0.25806452 0.74193548 1.         1.         1...   
Support Vector Machines  [0.58064516 0.83870968 1.         1.         1...   
k-Nearest Neighbors      [0.58064516 0.87096774 1.         1.         1...   

                              avg       std  
Logistic Regression      0.800000  0.288813  
Support Vector Machines  0.883871  0.163978  
k-Nearest Neighbors      0.890323  0.162703  


The k-Nearest Neighbors model performed the best with an average score of 0.890323. I think this is the case because the other 2 models make boundary lines which could completely inaccurate for a certain section of the data space, while kNN takes the spatial situation of each point into account when classifying them rather than brushing with broad strokes with a boundary. 

### Neural Networks

In [118]:
results = pd.DataFrame(columns=["scores", "avg", "std"])

def add_results(nn):
    nn.fit(X=x_train, y=y_train)
    row = []
    nn_scores = cross_val_score(nn, X=x, y=y, cv=kf)
    row.append(str(nn_scores))
    row.append(nn_scores.mean())
    row.append(nn_scores.std())
    results.loc[len(results)] = row

In [119]:
# Experiments with Neural Network.
#Default

nn = MLPClassifier(random_state=42)

add_results(nn)

C:\Users\raven\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\raven\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\raven\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
C:\Users\raven\A

In [120]:
#Increasing max_iterations
nn = MLPClassifier(max_iter=1000, random_state=42)

add_results(nn)

In [121]:
#Increasing learning rate
nn = MLPClassifier(learning_rate_init=1, random_state=42)

add_results(nn)

In [122]:
#Increasing learning rate and changing learning rate update schedule
nn = MLPClassifier(learning_rate="invscaling", learning_rate_init=1, random_state=42)

add_results(nn)

## Results and Conclusion for Neural Network Experiments

In [123]:
print(results)

                                              scores       avg       std
0  [0.51612903 0.74193548 1.         1.         1...  0.851613  0.195261
1  [0.74193548 0.93548387 0.93548387 1.         1...  0.922581  0.094819
2  [0.51612903 0.64516129 0.96774194 1.         1...  0.825806  0.204629
3  [0.51612903 0.64516129 0.96774194 1.         1...  0.825806  0.204629


I increased max iterations because when I did the default (avg score 0.851613), I got a warning that said that the optimization of the parameters was not able to be completed with the default amount of iterations, so I added more iterations and got way better results (avg score 0.922581), most likely because the parameters were able to finish optimizing. I then tried increasing the learning rate which allowed the model to be trained way faster (the first and second times were 0.9 seconds and 2.5 seconds respectively, with the third being 0.1 seconds) but the average score was worse than both of my previous trials (0.825806). I tried changing the learning rate update schedule to inverse scaling rather than constant, but everything, the time and the average score was the same. What made the biggest difference in accuracy was the number of iterations, what made the most difference in speed was the learning rate.